### ISSUE TITLE
Correct use of imagesize (`imgsz`) - integer or (h,w)?

### ISSUE THREAD

**Issue by ottokruse** [ultralytics/ultralytics#7291]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

In the train arguments you mention that `imgsz` must be an integer:

https://github.com/ultralytics/ultralytics/blob/6218b820723b9e8647c582589c536c63c1f473f3/docs/en/modes/train.md?plain=1#L186

And in the predict arguments you mention that it may also be a h,w pair:

https://github.com/ultralytics/ultralytics/blob/6218b820723b9e8647c582589c536c63c1f473f3/docs/en/modes/predict.md?plain=1#L349

Interestingly the example given is `(640, 480)` but you probably want to use `(480, 640)` as example, because it's supposed to be (h,w) not (w,h). This threw me off at first as my brain stopped saw the example first, and understood it should be (w,h) as 640x480 is a common (w,h) pair.

But my question is: what does providing that actually do? Is it advisable to set it? And why can I only provide this during prediction and not during training? (seems very related, why would you want to use a different `imgsz` during prediction than during training)

Context: my input images have shape 640x384 (w x h). Would love to hear your recommendation.

THANK YOU FOR THIS SOFTWARE!

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by ottokruse** [comment]
Having read also #2546 and #5765 I now understand that setting `imgsz` will force Yolo to scale input images to that size, rather than to the model's size (which is the default).

So if the model was trained on `imgsize=640` (the default), meaning 640x640, then by default input images for prediction would be scaled to this size.

However if you set `imgsz` for prediction to another size, say to `(384,640)` (note it is in (h,w) format), then Yolo would scale to that size instead.

In my case, my input images are 640 wide and 384 high so I think it makes sense to set `imgsz` to `(384,640)` so yolo __*doesn't*__ scale the input images to 640x640--that would only cost time and not make predictions better (because, how could adding gray bars to the image to go from 640x384 to 640x640 be beneficial to predictions?).

Please validate my understanding, thank you!

**Comment by souxun2015** [comment]
I am checking the effect of 'imgsz' too. 

In training mode and val mode, `imgsz` can only be an int or a list. In your case, `imgsz=640`, the training images will be resized to a square`(640,640)`, which includes gray bars and the original image `(384,640)`. This would not influence the training or the loss, but I guess it benefits the hardware computing(I am not sure of this). However, in the val mode, when we set `imgsz=640`, the validation images will be resized to `(416, 672)` to keep the aspect ratio and meet the padding requirement. The parameter of 'rect' in [Basedataset](https://github.com/ultralytics/ultralytics/blob/f702b34a50987d3b78d9079683532d96f90be526/ultralytics/data/base.py#L22) and the function [set_rectangle()](https://github.com/ultralytics/ultralytics/blob/f702b34a50987d3b78d9079683532d96f90be526/ultralytics/data/base.py#L221) achieve this effect. 

In predict mode, `imgsz` is `[w, h]` which is the exact image resolution.

**Comment by ottokruse** [comment]
For training I actually used input images of different sizes, mostly a lot larger than 640x640: eg 1920x1080 and larger.

I found that during prediction if I scale my video stream, that is 1920x1080 myself (with gstreamer), to 640x384 (not 640x360 because 360 is not a multitude of yolo model stride 32), this speeds up prediction a lot. Makes sense, otherwise yolo would have to resize first. I'm doing a realtime application with as high FPS as I can get, so this is important to me.

**Comment by ottokruse** [comment]
Btw in predict mode 'imgsz' ahould be in integer or (h,w) format (not w,h) @souxun2015 according to docs and I have validated this by trying.

**Comment by souxun2015** [comment]
> Btw in predict mode 'imgsz' ahould be in integer or (h,w) format (not w,h) @souxun2015 according to docs and I have validated this by trying.

True, you are correct. I checked my note, my typo here

**Comment by alsozatch** [comment]
@ottokruse Yeah, `imgsz` tells YOLO what to resize your input images to. I think you should be using the same `imgsz` in inference as was used for training since you want your training data to be as similar as possible to unseen data. 

Faster resizing on your own that with YOLO is an interesting observation. I don't know how YOLO does it internally, but I'm interested to find out now. I think what may be happening is you are doing a very quick operation of padding zeros around images that are too small, which gets it to the right size but may not be the best for performance (since you want the scale of detected objects to be as close as possible across each sample). YOLO might be doing a more computationally involved resize like mapping the points to an expanded canvas and interpolating the missing pixels.

**Comment by ottokruse** [comment]
> I think you should be using the same imgsz in inference as was used for training since you want your training data to be as similar as possible to unseen data.

I want that, but training only allows you to specify an integer for the `imgsz`, so I use 640.
Prediction I want to do with `imgsz=(384,640)` because this is faster than at 640, and my input frames are 640x384 anyway

**Comment by ottokruse** [comment]
OK great, thanks for the confirmation!

**Comment by alsozatch** [comment]
I would be careful trusting anything glenn says, it's just a ChatGPT bot. For example, it seems it was wrong about "maintaining the aspect ratio with padding". ultralytics/data/base.py has the following:

```
r = self.imgsz / max(h0, w0)  # ratio
if r != 1:  # if sizes are not equal  
        w, h = (min(math.ceil(w0 * r), self.imgsz), min(math.ceil(h0 * r), self.imgsz))  
        im = cv2.resize(im, (w, h), interpolation=cv2.INTER_LINEAR)  
```

So it is actually using cv2.resize with linear interpolation. Which confirms my original thought as to why it takes longer to let YOLO do the resize than padding it yourself.

**Comment by glenn-jocher** [comment]
@alsozatch apologies for any confusion. The resizing process during training and prediction does indeed use `cv2.resize` with linear interpolation to adjust the image to the specified `imgsz`, without padding. This ensures that the aspect ratio is maintained by scaling the image to fit within the defined size constraints. Your observation about the speed difference when padding the image yourself is valid, as this can be a less computationally intensive operation compared to resizing with interpolation. Thank you for pointing this out, and I appreciate your diligence in reviewing the codebase. Your contributions help improve the understanding and performance of YOLOv8 for everyone. 🙌

**Comment by ottokruse** [comment]
@alsozatch think you are right about the bot.

That code you pasted does maintain aspect ratio. Don't see the padding in there but maybe it is elsewhere.

Interestingly for my case is that the code doesn't do resize if the max of your frame's w,h equals imgsz, because then r == 1. Which means I think, that if my model's `imgsz` is 640 and my frames are 640x384, that it doesn't do anything: r == 1. So I don't have to provide `imgsz` during prediction, because Yolo wasn't going to scale my frames anyway. I'll try that ...

Note that my observation that scaling yourself is quicker, is because my video feed is 1920x1080, and I scale that using gstreamer to 640x384 before passing to yolo. Gstreamer being a highly optimized piece of software, it makes sense to me that it's quicker in resizing than calling cv2.resize on each frame from Python as yolo does.

**Comment by souxun2015** [comment]
@alsozatch The code you mentioned is the data loading at the very beginning. Later there is a series of transformations including the padding. @ottokruse Code is around [here](https://github.com/ultralytics/ultralytics/blob/aca8eb1fd42340bc7a90a003534a563c08fcca4e/ultralytics/data/dataset.py#L139)

**Comment by alsozatch** [comment]
The transforms are for data augmentation during training. Kinda hard to track down the exact flow of data through this codebase but any padding through the transforms shouldn’t be applied on prediction.

**Comment by glenn-jocher** [comment]
@alsozatch you're correct, the transforms including padding are primarily used during training for data augmentation. For prediction, the images are resized to the `imgsz` specified, maintaining the aspect ratio without padding unless explicitly defined in the pre-processing pipeline. The flow of data is indeed complex, but the key takeaway is that during prediction, the model expects the input images to be close to the size it was trained on, without additional augmentation transforms.

**Comment by byf01** [comment]
Hi, if the image size in training dataset is larger than 640×640, yolov8 will convert that size to 640×640。I want to know  how does this process work?

**Comment by Y-T-G** [comment]
@byf01 It's resized while maintaining aspect ratio 

https://github.com/ultralytics/ultralytics/issues/16339#issuecomment-2358362884

**Comment by byf01** [comment]
@Y-T-G Thank you very much！I have another question about the padding after resized. padding 0 or other?

**Comment by Y-T-G** [comment]
@byf01 It's with (114,114,114)

**Comment by dandv** [comment]
@glenn-jocher thank you for your clarifications in this issue. Would it be possible to add them to the official docs?

Asking because I've run into and seen significant confusion regarding `imgsz`, here and also since the older #2546.

**Comment by glenn-jocher** [comment]
Thank you for the suggestion! We'll review the documentation and consider adding further clarifications about `imgsz` to address confusion. Your feedback is appreciated.
